libs/langchain/langchain/llms/watsonxllm.py

```diff
@@ -0,0 +1,172 @@
+import logging
+import os
+from typing import Any, Dict, List, Mapping, Optional, Union
+
+from langchain.callbacks.manager import CallbackManagerForLLMRun
+from langchain.llms.base import LLM
+from langchain.llms.utils import enforce_stop_tokens
+from langchain.pydantic_v1 import Extra, root_validator
+from langchain.utils import get_from_dict_or_env
+
+logger = logging.getLogger(__name__)
+
+
+class WatsonxLLM(LLM):
+    """
+    IBM watsonx.ai large language models.
+
+    To use, you should have ``ibm_watson_machine_learning`` python package installed,
+    and the environment variable ``WATSONX_APIKEY`` set with your API key, or pass
+    it as a named parameter to the constructor.
+
+
+    Example:
+        .. code-block:: python
+
+            from langchain.llms import WatsonxLLM
+            llm = WatsonxLLM(
+                model_id='google/flan-ul2',
+                credentials = {"url": "https://us-south.ml.cloud.ibm.com",
+                                   "apikey": "*****"},
+                project_id="*****"
+            )
+    """
+
+    model_id: str = ""
+    """Type of model to use."""
+
+    project_id: str = ""
+    """ID of the Watson Studio project."""
+
+    space_id: str = ""
+    """ID of the Watson Studio space."""
+
+    credentials: dict = {}
```

Comment:
Could we unpack the credentials into fields and assign them an Optional[SecretStr] type

Reply:
Not mandatory, the only requirement is that when a user prints the object the secrets don't get printed (e.g., as might happen in an interactive environment like a jupyter notebook)

Other things that will be affected: when the model is serialized for tracing purposes, all credentials will need to be masked (including the URL)

Reply:
Credentials have been unpacked into Optional[SecretStr] type as you mentioned.